deepspeed/utils/timer.py

```diff
@@ -167,62 +171,71 @@ def __init__(
 
     def update_epoch_count(self):
         self.epoch_count += 1
-        self.local_step_count = 0
+        self.local_microstep_count = 0
 
     def _init_timer(self):
         self.initialized = True
 
     def start(self):
         self._init_timer()
         self.started = True
-        if self.total_step_count >= self.start_step:
+        if self.local_microstep_count >= self.start_step * self.gradient_accumulation_steps:
             torch.cuda.synchronize()
             self.start_time = time.time()
 
     def stop(self, report_speed=True):
         if not self.started:
             return
         self.started = False
-        self.total_step_count += 1
-        self.local_step_count += 1
-        if self.total_step_count > self.start_step:
+        self.local_microstep_count += 1
+        gradient_accumulation_boundary = (self.local_microstep_count %
+                                          self.gradient_accumulation_steps == 0)
+        if gradient_accumulation_boundary:
+            self.total_step_count += 1
+
+        if self.local_microstep_count > self.start_step * self.gradient_accumulation_steps:
```

Comment:
`if self.global_step_count >= self.start_step`

Reply:
This, I believe, won't work as we need to accumulate the duration of the first micro steps before first full step after `self.start_step`

Reply:
My idea is to have the client explicitly control when micro_steps and global_steps are updated so that the timer does not have to reason about or track the relationship of these two. Specifically, 
1. micro_steps is updated when client calls timer.stop()
2. global_steps is updated when client calls timer.stop(...,global_step=True,...) 

Does that make sense?

Reply:
I can change it to `if self.start_time > 0:` if there are no objections.